nearcore/src/lib.rs

```diff
@@ -498,3 +499,90 @@ pub fn start_with_config_and_synchronization(
         arbiters: vec![client_arbiter_handle, arbiter.handle()],
     })
 }
+
+pub fn recompress_storage(home_dir: &Path, dst_dir: &Path) -> anyhow::Result<()> {
+    use strum::{EnumCount, IntoEnumIterator};
+
+    // We’re configuring each RocksDB to use 512 file descriptors.  Make sure we
+    // can open that many by ensuring nofile limit is large enough to give us
+    // some room to spare over 1024 file descriptors.
+    let (soft, hard) = rlimit::Resource::NOFILE
+        .get()
+        .map_err(|err| anyhow::anyhow!("getrlimit: NOFILE: {}", err))?;
+    if soft < 3 * 512 {
+        rlimit::Resource::NOFILE
+            .set(3 * 512, hard)
+            .map_err(|err| anyhow::anyhow!("setrlimit: NOFILE: {}", err))?;
+    }
+
```

Comment:
If 512 is a config value why do we hard code it here?

Reply:
It’s hard coded in `db.rs` in `rocksdb_options`:

```rust
/// DB level options
fn rocksdb_options() -> Options {
    let mut opts = Options::default();

    set_compression_options(&mut opts);
    opts.create_missing_column_families(true);
    opts.create_if_missing(true);
    opts.set_use_fsync(false);
    opts.set_max_open_files(512);  // <------ here
    opts.set_keep_log_file_num(1);
    opts.set_bytes_per_sync(bytesize::MIB);
    opts.set_write_buffer_size(256 * bytesize::MIB as usize);
    opts.set_max_bytes_for_level_base(256 * bytesize::MIB);
    // ...

    opts
}
```

It should be an option actually, but that’s a separate story.

Reply:
> As to why not make it a constant I honestly intend for the recompress storage command to be temporary (as in just couple of releases to let anyone who cares recompress) so admittedly I am cutting corners a little.

What is the reason for not keeping it permanent? Is it because recompression would happen eventually anyways?

Reply:
The recompression would happen eventually but I don’t think that would happen in reasonable time.  I haven’t tested it and could be wrong but to recompress the whole database all of the SST files would need to be touched and that means a lot of compactions through all the levels.

It’s more that I hope that in six months there will be no use for this code. Anyone who starts a new archival node in six months will use the database with new compression (our snapshots will be recompressed and their node will use the new compression). And anyone who is running archival node will use this command to recompress their storage (or start anew from our snapshot depending on their trust requirements).

And even more than that, this code to me feels like a hack prepared just for the sake of the compression migration rather than a command that belongs to neard.